### DISCUSSION TITLE
Apostrophe 3.0 Stable is here! 🎊

### DISCUSSION THREAD

**Discussion by agilbert** [apostrophecms/apostrophe#3176] [Release Notes]
Hi everyone, it's a big day for Apostrophe! **Today we are so very pleased to share our 3.0 Stable release of ApostropheCMS with you.** This release represents years of work from our core team as well as contributors from around the world.

If you're reading this, you are probably already familiar with A3 and the progress we've been making over the last few months. We'll have a more robust announcement published tomorrow on apostrophecms.com and our mailing list. In the meantime, here are some links to jump right in:

* [Check out the live demo](http://demo.apostrophecms.org/)
* [Dig into the new documentation](http://v3.docs.apostrophecms.org/)
* [Start a project with A3](https://v3.docs.apostrophecms.org/guide/setting-up.html#system-requirements)
* [Reach out to us for support](https://apostrophecms.com/contact-us)

A special thank you to everyone that's been active here in GitHub and on [our Discord](http://chat.apostrophecms.com/) as we've been working towards this milestone. This is merely the beginning of a huge new era for Apostrophe and we can't wait for future collaborations with the community around A3.

Read on for the detailed changelog as well as other modules that were published this week.

### Breaks

* Previously our `a3-boilerplate` project came with a webpack build that pushed code to the `ui/public` folder of an `asset` module. Now the webpack build is not needed because Apostrophe takes care of compiling `ui/src` for us. This is good! However, **if you are transitioning your project to this new strategy, you will need to remove the `modules/asset/ui/public` folder from your project manually** to ensure that webpack-generated code originally intended for webpack-dev-server does not fail with a `publicPath` error in the console.
* The `CORE_DEV=1` environment setting has been changed to `APOS_DEV=1` because it is appropriate for anyone who is actively developing custom Apostrophe admin UI using `ui/apos` folders in their own modules.
* Apostrophe now uses Dart Sass, aka the `sass` npm module. The `node-sass` npm module has been deprecated by its authors for some time now. Most existing projects will be unaffected, but those writing their own Apostrophe UI components will need to change any `/deep/` selectors to `::v-deep` and consider making other Dart Sass updates as well. For more information see the [Dart Sass documentation](https://sass-lang.com/dart-sass). Those embracing the new `ui/src` feature should also bear in mind that Dart Sass is being used.

### Changes

* Relationship ids are now stored as aposDocIds (without the locale and mode part). The appropriate locale and mode are known from the request. This allows easy comparison and copying of these properties across locales and fixes a bug with reverse relationships when publishing documents. A migration has been added to take care of this conversion on first startup.
- The `attachment` field type now correctly limits file uploads by file type when using the `fileGroup` field option.
- Uploading SVG files is permitted in the Media Library by default.

### Adds

- Apostrophe now enables you to ship frontend JavaScript and Sass (using the SCSS syntax) without your own webpack configuration.
- Any module may contain modern JavaScript in a `ui/src/index.js` file, which may use `import` to bring in other files in the standard way. Note that **`ui/src/index.js must export a function`**. These functions are called for you in the order modules are initialized.
- Any module may contain a Sass (SCSS) stylesheet in a `ui/src/index.scss` file, which may also import other Sass (SCSS) files.
- Any project that requires IE11 support for `ui/src` JavaScript code can enable it by setting the `es5: true` option to the `@apostrophecms/asset` module. Apostrophe produces separate builds for IE11 and modern browsers, so there is no loss of performance in modern browsers. Code is automatically compiled for IE11 using `babel` and missing language features are polyfilled using `core-js` so you can use promises, `async/await` and other standard modern JavaScript features.
- `ui/public` is still available for raw JavaScript and CSS files that should be pushed *as-is* to the browser. The best use of this feature is to deliver the output of your own custom webpack build, if you have one.
- Adds browser-side `editMode` flag that tracks the state of the current view (edit or preview), located at `window.apos.adminBar.editMode`.
- Support for automatic inline style attribute sanitization for Rich Text widgets.
- Adds text align controls for Rich Text widgets. The following tools are now supported as part of a rich text widget's `toolbar` property:
-- `alignLeft`
-- `alignRight`
-- `alignCenter`
-- `alignJustify`
- `@apostrophecms/express` module now supports the `trustProxy: true` option, allowing your reverse proxy server (such as nginx) to pass on the original hostname, protocol and client IP address.

### Fixes

* Unit tests passing again. Temporarily disabled npm audit checks as a source of critical failures owing to upstream issues with third-party packages which are not actually a concern in our use case.
* Fixed issues with the query builder code for relationships. These issues were introduced in beta 3 but did not break typical applications, except for displaying distinct choices for existing values of a relationship field.
* Checkbox field types can now be used as conditional fields.
* Tracks references to attachments correctly, and introduces a migration to address any attachments previously tracked as part of documents that merely have a relationship to the proper document, i.e. pages containing widgets that reference an image piece.
* Tracks the "previously published" version of a document as a legitimate reference to any attachments, so that they are not discarded and can be brought back as expected if "Undo Publish" is clicked.
* Reverse relationships work properly for published documents.
* Relationship subfields are now loaded properly when `reverseOf` is used.
* "Discard Draft" is available when appropriate in "Manage Pages" and "Manage Pieces."
* "Discard Draft" disables the "Submit Updates" button when working as a contributor.
* Relationship subfields can now be edited when selecting in the full "manage view" browser, as well as in the compact relationship field view which worked previously.
* Relationship subfields now respect the `def` property.
* Relationship subfields are restored if you deselect a document and then reselect it within a single editing experience, i.e. accidentally deselect and immediately reselect, for instance.
* A console warning when editing subfields for a new relationship was fixed.
* Field type `color`'s `format` option moved out of the UI options and into the general options object. Supported formats are "rgb", "prgb", "hex6", "hex3", "hex8", "name", "hsl", "hsv". Pass the `format` string like:
```js
myColorField: {
  type: 'color',
  label: 'My Color',
  options: {
    format: 'hsl'
  }
}
```
* Restored Vue dependency to using semantic versioning now that Vue 2.6.14 has been released with a fix for the bug that required us to pin 2.6.12.
* Nunjucks template loader is fully compatible with Linux in a development environment.
* Improved template performance by reusing template loaders.
* `min` and `max` work properly for both string-like and number-like fields.
* Negative numbers, leading minus and plus signs, and trailing periods are accepted in the right ways by appropriate field types.
* If a user is inadvertently inserted with no password, set a random password on the backend for safety. In tests it appears that login with a blank password was already forbidden, but this provides an additional level of certainty.
* `data.page` and `data.contextOptions` are now available in `widget.html` templates in most cases. Specifically, they are available when loading the page, (2) when a widget has just been inserted on the page, and (3) when a widget has just been edited and saved back to the page. However, bear in mind that these parameters are never available when a widget is being edited "out of context" via "Page Settings", via the "Edit Piece" dialog box, via a dialog box for a parent widget, etc. Your templates should be written to tolerate the absence of these parameters.
* Double slashes in the slug cannot be used to trick Apostrophe into serving as an open redirect (fix ported to 3.x from 2.92.0).
* The global doc respects the `def` property of schema fields when first inserted at site creation time.
* Fixed fragment keyword arguments being available when not a part of the fragment signature.

## Apostrophe 3.x extensions

[cli 3.0.0](https://github.com/apostrophecms/cli)

* The initial build of the overhauled ApostropheCMS CLI. Uses the 3.0.0 major version number as this is very much an advanced version of the apostrophe-cli package (currently at 2.x.x), but moved to a new package name for logistical reasons.

[seo 1.0.0](https://github.com/apostrophecms/seo)

* Initial Apostrophe 3 Release

[open-graph 1.0.0](https://github.com/apostrophecms/open-graph)

* Initial Apostrophe 3 Release

## Apostrophe 2.x

[apostrophe-workflow 2.39.1](https://github.com/apostrophecms/apostrophe-workflow)

* Fixes duplicate key errors in sync page tree task. For live pages not already updated with the same path.

[apostrophe-pieces-export 2.4.1](https://github.com/apostrophecms/apostrophe-pieces-export)

* Fix Include Workflow filter

[apostrophe-forms-google-address-field-widgets 3.0.0](https://github.com/apostrophecms/apostrophe-forms-google-address-field-widgets)

* Adds labels to splitted address fields

## Other modules

[mechanic 1.7.0](https://github.com/apostrophecms/mechanic)

* Support for path-specific backends, for microservices and the like.

**Comment by fredrikekelund** [comment]
Congratulations on the big release! 🎉 Haven't had a chance to play around much with the new version yet, but I'm looking forward to do so

**Reply by gitvadim** [reply]
I've been looking forward to A3 ever since it was announced. A few days in with it and so far I'm loving it. Everything is so much more making sense compared to A2.

**Reply by abea** [reply]
So glad to hear that!

**Reply by myovchev** [reply]
I'm building on it 6+ months now and it feels like Home.
Keep it up team Apostrophe.

**Comment by myovchev** [comment]
`mechanic` is a nice little surprise! I love how it sounds and my team will definitely play around as we do all of the described in the docs mostly manually. Not to mention it is adaptable to any custom needs (and it uses nunjucks!!!) :D

**Reply by abea** [reply]
For those of us who aren't nginx pros it is extremely useful.

**Reply by myovchev** [reply]
Well, the reaction of my nginx pro teammate was: "Wow, they did a console cPanel" :D

**Comment by boutell** [comment]
Haha it's true!

One nice touch: "mechanic list" produces mechanic commands that can be used
to reproduce the same configuration on a new server.

On Fri, Jun 25, 2021 at 1:57 PM Miro Yovchev ***@***.***>
wrote:

> Well, the reaction of my nginx pro teammate was: "Wow, they did a console
> cPanel" :D
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apostrophecms/apostrophe/discussions/3176#discussioncomment-922325>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAH27JMPUDWS6IQFMCOGLTTUS7QFANCNFSM462KBOOQ>
> .
>


-- 

THOMAS BOUTELL | CHIEF TECHNOLOGY OFFICER
APOSTROPHECMS | apostrophecms.com | he/him/his

**Reply by myovchev** [reply]
Oh, this was the most exciting feature I've spotted on first glance!
